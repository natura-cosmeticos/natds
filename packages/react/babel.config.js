const path = require('path')

const defaultAlias = {
  '@naturacosmeticos/natds-react': path.resolve(__dirname, './src/components')
}

module.exports = {
  plugins: [
    ['babel-plugin-module-resolver', { alias: defaultAlias }]
  ],
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    '@babel/preset-typescript',
    '@babel/preset-react'
  ]
}
