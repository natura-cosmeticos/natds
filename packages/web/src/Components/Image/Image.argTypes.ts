import { ImageProps } from "./Image.props";
import { sizePropArgType } from "../../../.storybook/argTypes/sizePropArgType";

export const argTypes : Record<keyof ImageProps | string, unknown> = {
  alt: {
    control: {
      type: "text",
    },
  },
  className: {
    control: {
      type: "text",
    },
  },
  draggable: {
    control: {
      options: [
        true,
        false,
        "auto",
      ],
      type: "inline-radio",
    },
    description: `
The draggable property specifies whether an element is draggable or not.

Images are draggable by default.

If \`true\` the element is draggable.
If \`false\` the element is not draggable.
If \`"auto"\` uses the default behavior of the browser.

`,
    table: {
      defaultValue: {
        summary: "auto",
      },
      type: {
        summary: "true | false | \"auto\"",
      },
    },
  },
  disableSelection: {},
  height: {
    control: {
      type: "number",
    },
    table: sizePropArgType.table,
  },
  maxHeight: sizePropArgType,
  maxWidth: sizePropArgType,
  fallback: {
    control: {
      type: "text",
    },
  },
  src: {
    control: {
      type: "text",
    },
  },
  state: {},
  style: {
    control: {
      type: "object",
    },
    table: {
      defaultValue: {
        summary: "{}",
      },
    },
  },
  width: {
    control: {
      type: "number",
    },
    table: sizePropArgType.table,
  },
  radius: {},
};
