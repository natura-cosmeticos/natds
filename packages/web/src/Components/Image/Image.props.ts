import * as React from "react";
import { SizePropValue } from "../../hooks/useSizeStyleProp";

export type ImageVariant = "standard" | "highlight";
export type ImageFallback = "product" | React.ImgHTMLAttributes<HTMLImageElement>["src"];

export interface ImageProps extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, "height" | "width"> {

  /**
   * Image alternative text.
   *
   * @required
   * @type string
   */
  alt: React.ImgHTMLAttributes<HTMLImageElement>["alt"];

  /**
   * `className` property which is applied to the root element.
   *
   * @optional
   * @type string
   */
  className?: React.HTMLAttributes<HTMLImageElement>["className"];

  /**
   * Prevents image selection.
   *
   * If `false`, image can be selected by the user if the browser allows it.
   * If `true`, image selection is disabled if the browser supports `user-select` CSS property.
   *
   * @optional
   * @type boolean
   */
  disableSelection?: boolean;

  /**
   * Add fallback image.
   *
   * When image loads with error apply fallback. If fallback is equal to "product" then apply product-empty image.
   *
   * @optional
   * @type "product" | string
   */
  fallback?: ImageFallback;

  /**
   * Sets the image height.
   *
   * Number of pixels or size token for image height.
   * You can set `"auto"` if you want to let the image determine the height.
   *
   * @optional
   * @type SizePropValue
   */
  height?: SizePropValue;

  /**
   * Sets the maximum image height.
   *
   * Prevents the `height` property value from becoming greater than the value indicated by `maxHeight`.
   *
   * @optional
   * @type SizePropValue
   */
  maxHeight?: SizePropValue;

  /**
   * Sets the maximum image width.
   *
   * Prevents the `width` property value from becoming greater than the value indicated by `maxWidth`.
   *
   * @optional
   * @type SizePropValue
   */
  maxWidth?: SizePropValue;

  /**
   * Sets the image border-radius.
   *
   * If `true`, image border-radius is set to 4px.
   * If `false`, image border-radius is set to none.
   *
   * @optional
   * @type boolean
   */
  radius?: boolean;

  /**
   * An alias for `image` property.
   *
   * @optional
   * @type string
   */
  src?: React.ImgHTMLAttributes<HTMLImageElement>["src"];

  /**
   * Add overlay to the image.
   *
   * If `true`, add image overlay.
   * If `false`, no overlay.
   *
   * @optional
   * @type boolean
   */
  state?: boolean;

  /**
   * Override with inline-styles
   *
   * This inline-style approach is recommended only for dynamic style properties.
   * The CSS properties will override the style of `Image` component (root).
   *
   * Inline styles take precedence over the regular CSS.
   *
   * @optional
   * @type React.CSSProperties
   */
  style?: React.CSSProperties

  /**
   * Sets the image width.
   *
   * Number of pixels or size token for image width.
   * You can set `"auto"` if you want to let the image determine the width.
   *
   * @optional
   * @type SizePropValue
   */
  width?: SizePropValue;

}
