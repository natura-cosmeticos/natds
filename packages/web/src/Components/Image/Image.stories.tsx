// eslint-disable-next-line no-use-before-define
import * as React from "react";
import { Meta, Story } from "@storybook/react";
import { Image } from "./Image";
import { ImageProps } from "./Image.props";
import { argTypes } from "./Image.argTypes";

export default {
  argTypes,
  component: Image,
  title: "Components/Image",
} as Meta;

const Template : Story<ImageProps> = (args: ImageProps) => <Image {...args} />;

export const Playground : Story<ImageProps> = Template.bind({});
Playground.args = {
  alt: "Nat Natura holding a cat",
  src: "https://pbs.twimg.com/profile_images/1308179129267781632/wLDnEiOy_400x400.jpg",
};

export const WithOverlay : Story<ImageProps> = Template.bind({});
WithOverlay.args = {
  alt: "Nat Natura holding a cat",
  src: "https://pbs.twimg.com/profile_images/1308179129267781632/wLDnEiOy_400x400.jpg",
  state: true,
};

export const WithRadius : Story<ImageProps> = Template.bind({});
WithRadius.args = {
  alt: "Nat Natura holding a cat",
  src: "https://pbs.twimg.com/profile_images/1308179129267781632/wLDnEiOy_400x400.jpg",
  radius: true,
};

export const Fallback : Story<ImageProps> = Template.bind({});
Fallback.args = {
  alt: "Nat Natura holding a cat",
  src: "https://pbs.twimg.com/profile_images/308179129267781632/wLDnEiOy_400x400.jpg",
  fallback: "https://zeroheight.com/uploads/CKp-xkMzUziuVtBy4yXxyg.jpg",
  maxWidth: "veryHuge",
};

export const ProductFallback : Story<ImageProps> = Template.bind({});
ProductFallback.args = {
  alt: "Nat Natura holding a cat",
  src: "https://pbs.twimg.com/profile_images/308179129267781632/wLDnEiOy_400x400.jpg",
  fallback: "product",
};
