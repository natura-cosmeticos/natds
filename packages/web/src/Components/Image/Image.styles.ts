import makeStyles from '@material-ui/core/styles/makeStyles'
import createStyles from '@material-ui/core/styles/createStyles'
import { ImageProps } from './Image.props'
import { IThemeWeb } from '../../Themes'
import { getSizeStyleProp } from '../../hooks/useSizeStyleProp'

export const useStyles = makeStyles((theme: IThemeWeb) => createStyles({
  root: ({
    disableSelection, height, maxHeight, maxWidth, radius, width
  }: ImageProps) => ({
    borderRadius: radius ? theme.shape?.borderRadius : 'none',
    height: getSizeStyleProp({
      theme,
      value: height
    }),
    maxHeight: getSizeStyleProp({
      theme,
      value: maxHeight
    }),
    maxWidth: getSizeStyleProp({
      theme,
      value: maxWidth
    }),
    userSelect: disableSelection ? 'none' : 'auto',
    width: getSizeStyleProp({
      theme,
      value: width
    })
  }),
  container: ({ disableSelection, radius }: ImageProps) => ({
    borderRadius: radius ? theme.shape?.borderRadius : 'none',
    display: 'inline-block',
    lineHeight: '0px',
    position: 'relative',
    userSelect: disableSelection ? 'none' : 'auto'
  }),
  overlay: {
    backgroundColor: 'rgba(0, 0, 0, 0.55)',
    borderRadius: ({ radius }: ImageProps) => (radius ? theme.shape?.borderRadius : 'none'),
    position: 'absolute',
    left: 0,
    top: 0,
    bottom: 0,
    right: 0,
    opacity: '0',
    '&:hover': {
      opacity: '1'
    },
    transition: 'opacity .3s'
  }
}), { name: 'NatDSImage' })

export default useStyles
