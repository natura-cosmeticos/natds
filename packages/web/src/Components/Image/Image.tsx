// eslint-disable-next-line no-use-before-define
import * as React from "react";
import useTheme from "@material-ui/core/styles/useTheme";
import clsx from "clsx";
import { ImageProps } from "./Image.props";
import { useStyles } from "./Image.styles";
import { IThemeWeb } from "../../Themes";
import { getSizeStyleProp } from "../../hooks/useSizeStyleProp";

export { ImageProps } from "./Image.props";

export const Image = React.forwardRef<HTMLImageElement, ImageProps>(
  // eslint-disable-next-line complexity, max-statements
  (props: ImageProps, ref) => {
    const {
      alt, className, draggable, disableSelection, fallback, height, maxHeight, maxWidth, radius, state, style, src, width, ...otherProps
    } = props;

    const classes = useStyles({
      alt, disableSelection, height, maxHeight, maxWidth, radius, state, src, width,
    });

    const theme : IThemeWeb = useTheme();

    const heightAttribute = getSizeStyleProp({
      theme,
      value: height,
    });
    const widthAttribute = getSizeStyleProp({
      theme,
      value: width,
    });

    const [imageSrc, setImageSrc] = React.useState({ src, error: false });

    const onError = () => {
      if (!imageSrc.error) {
        setImageSrc({ src: fallback, error: true });

        return null;
      }

      return null;
    };

    const heightAttr = heightAttribute === "auto" ? "" : heightAttribute;
    const widthAttr = widthAttribute === "auto" ? "" : widthAttribute;

    if (state) {
      return (
        <div className={classes.container}>
          <img
            alt={alt}
            className={clsx(classes.root, className)}
            draggable={draggable}
            height={heightAttr}
            onError={onError}
            ref={ref}
            src={imageSrc.src}
            style={style}
            width={widthAttr}
            {...otherProps}
          />
          <div className={classes.overlay}></div>
        </div>
      );
    }

    return (
      <img
        alt={alt}
        className={clsx(classes.root, className)}
        draggable={draggable}
        height={heightAttr}
        onError={onError}
        ref={ref}
        src={imageSrc.src}
        style={style}
        width={widthAttr}
        {...otherProps}
      />
    );
  },
);

Image.displayName = "Image";

export default Image;
