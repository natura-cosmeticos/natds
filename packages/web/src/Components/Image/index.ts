import Image from './Image'

export { ImageProps } from './Image.props'
export { Image }
export default Image
