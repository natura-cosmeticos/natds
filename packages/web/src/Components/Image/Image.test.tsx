// eslint-disable-next-line no-use-before-define
import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import Image from "./Image";

describe("Image component", () => {
  let testRenderer : TestRenderer.ReactTestRenderer = {} as TestRenderer.ReactTestRenderer;

  describe("when only required properties are provided", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} />);

      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when disableSelection property is explicitly set to true", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} disableSelection={true} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when disableSelection property is explicitly set to false", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} disableSelection={false} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when height is set to auto", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} height={"auto"} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when width is set to auto", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} width={"auto"} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when state is true", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} state={true} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when fallback property exist", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} fallback={"https://zeroheight.com/uploads/CKp-xkMzUziuVtBy4yXxyg.jpg"} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when radius is true", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} radius={true} />);
      expect(testRenderer).toMatchSnapshot();
    });
  });
  describe("when error", () => {
    it("should render correctly", () => {
      testRenderer = TestRenderer.create(<Image alt={"Image alt text"} src="https://pbs.twimg.com/profile_images/308179129267781632/wLDnEiOy_400x400.jpg" />);

      const img = testRenderer.root;
      const imgError = img.findByType("img");

      TestRenderer.act(() => {
        imgError.props.onError();
      });
      TestRenderer.act(() => {
        imgError.props.onError();
      });

      expect(testRenderer).toMatchSnapshot();
    });
  });
});
