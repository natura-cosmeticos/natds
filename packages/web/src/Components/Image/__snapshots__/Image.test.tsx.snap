// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component when disableSelection property is explicitly set to false should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-10"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when disableSelection property is explicitly set to true should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-7"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when error should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-28"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when fallback property exist should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-22"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when height is set to auto should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-13"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when only required properties are provided should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-4"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when radius is true should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-25"
  height=""
  onError={[Function]}
  width=""
/>
`;

exports[`Image component when state is true should render correctly 1`] = `
<div
  className="NatDSImage-container-2 NatDSImage-container-20"
>
  <img
    alt="Image alt text"
    className="NatDSImage-root-1 NatDSImage-root-19"
    height=""
    onError={[Function]}
    width=""
  />
  <div
    className="NatDSImage-overlay-3 NatDSImage-overlay-21"
  />
</div>
`;

exports[`Image component when width is set to auto should render correctly 1`] = `
<img
  alt="Image alt text"
  className="NatDSImage-root-1 NatDSImage-root-16"
  height=""
  onError={[Function]}
  width=""
/>
`;
